Test teardown fails with `ERROR .hypothesis/tmp - FileNotFoundError: [Errno 2]`

In our [CI Job](https://github.com/kkrt-labs/keth/actions/runs/12269297723/job/34232619504) in which we use `pytest` and `hypothesis`, we run our tests with           [uv run pytest -n logical --junitxml=junit.xml -o junit_family=legacy](https://github.com/kkrt-labs/keth/blob/54912a7c64448fc92a6745321aebce7f9cfb6a0b/.github/workflows/python_tests.yml#L32)

After 100% of tests are run, the process fails with error:

```
ERROR .hypothesis/tmp - FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/keth/keth/cairo/.hypothesis/tmp'
```

I'm not sure exactly where this could come from, as it isn't always systematic.

The error message 

```
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/runner/work/keth/keth/cairo
configfile: pyproject.toml
plugins: asyncio-0.24.0, typeguard-2.13.3, hypothesis-6.122.3, xdist-3.6.1, profiling-1.8.1
asyncio: mode=strict, default_loop_scope=session
created: 64/64 workers
64 workers [525 items]

........................................................................ [ 13%]
........................................................................ [ 27%]
........................................................................ [ 41%]
........................................................................ [ 54%]
........................................................................ [ 68%]
........................................................................ [ 82%]
........................................................................ [ 96%]
.....................                                                    [100%]
==================================== ERRORS ====================================
_______________________ ERROR collecting .hypothesis/tmp _______________________
../.venv/lib/python3.10/site-packages/_pytest/runner.py:341: in from_call
    result: TResult | None = func()
../.venv/lib/python3.10/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
../.venv/lib/python3.10/site-packages/_pytest/main.py:523: in collect
    for direntry in scandir(self.path):
../.venv/lib/python3.10/site-packages/_pytest/pathlib.py:950: in scandir
    with os.scandir(path) as s:
E   FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/keth/keth/cairo/.hypothesis/tmp'
```

suggests that the error is in `pytest_make_collect_report`  and eventually in this classe's `collect`

```
@final
class Dir(nodes.Directory):
    """Collector of files in a file system directory.

    .. versionadded:: 8.0

    .. note::

        Python directories with an `__init__.py` file are instead collected by
        :class:`~pytest.Package` by default. Both are :class:`~pytest.Directory`
        collectors.
    """

    @classmethod
    def from_parent(  # type: ignore[override]
        cls,
        parent: nodes.Collector,
        *,
        path: Path,
    ) -> Self:
        """The public constructor.

        :param parent: The parent collector of this Dir.
        :param path: The directory's path.
        :type path: pathlib.Path
        """
        return super().from_parent(parent=parent, path=path)

    def collect(self) -> Iterable[nodes.Item | nodes.Collector]:
        config = self.config
        col: nodes.Collector | None
        cols: Sequence[nodes.Collector]
        ihook = self.ihook
        for direntry in scandir(self.path):
            if direntry.is_dir():
                path = Path(direntry.path)
                if not self.session.isinitpath(path, with_parents=True):
                    if ihook.pytest_ignore_collect(collection_path=path, config=config):
                        continue
                col = ihook.pytest_collect_directory(path=path, parent=self)
                if col is not None:
                    yield col

            elif direntry.is_file():
                path = Path(direntry.path)
                if not self.session.isinitpath(path):
                    if ihook.pytest_ignore_collect(collection_path=path, config=config):
                        continue
                cols = ihook.pytest_collect_file(file_path=path, parent=self)
                yield from cols
```

what is this `.hypothesis/tmp` file actually?

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,8 @@
+RELEASE_TYPE: patch
+
+Our pytest plugin now emits a warning if you set Pytest's ``norecursedirs``
+config option in such a way that the ``.hypothesis`` directory would be
+searched for tests.  This reliably indicates that you've made a mistake
+which slows down test collection, usually assuming that your configuration
+extends the set of ignored patterns when it actually replaces them.
+(:issue:`4200`)

--- a/hypothesis-python/src/_hypothesis_pytestplugin.py
+++ b/hypothesis-python/src/_hypothesis_pytestplugin.py
@@ -24,6 +24,7 @@ import json
 import os
 import sys
 import warnings
+from fnmatch import fnmatch
 from inspect import signature
 
 import _hypothesis_globals
@@ -443,6 +444,26 @@ else:
 
     _orig_call = fixtures.FixtureFunctionMarker.__call__
     fixtures.FixtureFunctionMarker.__call__ = _ban_given_call  # type: ignore
+
+    if int(pytest.__version__.split(".")[0]) >= 7:  # pragma: no branch
+        # Hook has had this signature since Pytest 7.0, so skip on older versions
+
+        def pytest_ignore_collect(collection_path, config):
+            # Detect, warn about, and mititgate certain misconfigurations;
+            # this is mostly educational but can also speed up collection.
+            if (
+                (name := collection_path.name) == ".hypothesis"
+                and collection_path.is_dir()
+                and not any(fnmatch(name, p) for p in config.getini("norecursedirs"))
+            ):
+                warnings.warn(
+                    "Skipping collection of '.hypothesis' directory - this usually "
+                    "means you've explicitly set the `norecursedirs` pytest config "
+                    "option, replacing rather than extending the default ignores.",
+                    stacklevel=1,
+                )
+                return True
+            return None  # let other hooks decide
 
 
 def load():

--- a/hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py
+++ b/hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py
@@ -269,21 +269,6 @@ class LazySequenceCopy:
         return i
 
 
-def clamp(lower: float, value: float, upper: float) -> float:
-    """Given a value and lower/upper bounds, 'clamp' the value so that
-    it satisfies lower <= value <= upper."""
-    # this seems pointless (and is for integers), but handles the -0.0/0.0 case.
-    # clamp(-1, 0.0, -0.0) violates the bounds by returning 0.0, since min(0.0, -0.0)
-    # takes the first value of 0.0.
-    if lower == value == upper:
-        return value
-    if value == lower:
-        return lower
-    if value == upper:
-        return upper
-    return max(lower, min(value, upper))
-
-
 def swap(ls: LazySequenceCopy, i: int, j: int) -> None:
     """Swap the elements ls[i], ls[j]."""
     if i == j:

--- a/hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py
+++ b/hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py
@@ -275,6 +275,8 @@ def clamp(lower: float, value: float, upper: float) -> float:
     # this seems pointless (and is for integers), but handles the -0.0/0.0 case.
     # clamp(-1, 0.0, -0.0) violates the bounds by returning 0.0, since min(0.0, -0.0)
     # takes the first value of 0.0.
+    if lower == value == upper:
+        return value
     if value == lower:
         return lower
     if value == upper:

--- a/hypothesis-python/src/hypothesis/internal/floats.py
+++ b/hypothesis-python/src/hypothesis/internal/floats.py
@@ -13,8 +13,6 @@ import struct
 from sys import float_info
 from typing import TYPE_CHECKING, Callable, Literal, SupportsFloat, Union
 
-from hypothesis.internal.conjecture.junkdrawer import clamp
-
 if TYPE_CHECKING:
     from typing import TypeAlias
 else:
@@ -206,6 +204,17 @@ def sign_aware_lte(x: float, y: float) -> bool:
         return math.copysign(1.0, x) <= math.copysign(1.0, y)
     else:
         return x <= y
+
+
+def clamp(lower: float, value: float, upper: float) -> float:
+    """Given a value and lower/upper bounds, 'clamp' the value so that
+    it satisfies lower <= value <= upper."""
+    # this seems pointless (and is for integers), but handles the -0.0/0.0 case.
+    if sign_aware_lte(value, lower):
+        return lower
+    if sign_aware_lte(upper, value):
+        return upper
+    return value
 
 
 SMALLEST_SUBNORMAL = next_up(0.0)

--- a/hypothesis-python/src/hypothesis/internal/floats.py
+++ b/hypothesis-python/src/hypothesis/internal/floats.py
@@ -208,11 +208,11 @@ def sign_aware_lte(x: float, y: float) -> bool:
 
 def clamp(lower: float, value: float, upper: float) -> float:
     """Given a value and lower/upper bounds, 'clamp' the value so that
-    it satisfies lower <= value <= upper."""
+    it satisfies lower <= value <= upper.  NaN is mapped to lower."""
     # this seems pointless (and is for integers), but handles the -0.0/0.0 case.
-    if sign_aware_lte(value, lower):
+    if not sign_aware_lte(lower, value):
         return lower
-    if sign_aware_lte(upper, value):
+    if not sign_aware_lte(value, upper):
         return upper
     return value
 

--- a/hypothesis-python/src/hypothesis/strategies/_internal/utils.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/utils.py
@@ -17,8 +17,7 @@ import attr
 
 from hypothesis.internal.cache import LRUReusedCache
 from hypothesis.internal.compat import dataclass_asdict
-from hypothesis.internal.conjecture.junkdrawer import clamp
-from hypothesis.internal.floats import float_to_int
+from hypothesis.internal.floats import clamp, float_to_int
 from hypothesis.internal.reflection import proxies
 from hypothesis.vendor.pretty import pretty
 

--- a/hypothesis-python/tests/conjecture/test_junkdrawer.py
+++ b/hypothesis-python/tests/conjecture/test_junkdrawer.py
@@ -20,13 +20,12 @@ from hypothesis.internal.conjecture.junkdrawer import (
     NotFound,
     SelfOrganisingList,
     binary_search,
-    clamp,
     endswith,
     replace_all,
     stack_depth_of_caller,
     startswith,
 )
-from hypothesis.internal.floats import float_to_int, sign_aware_lte
+from hypothesis.internal.floats import clamp, float_to_int, sign_aware_lte
 
 
 def test_out_of_range():

--- a/hypothesis-python/tests/conjecture/test_junkdrawer.py
+++ b/hypothesis-python/tests/conjecture/test_junkdrawer.py
@@ -77,6 +77,8 @@ def clamp_inputs(draw):
 @example((5, 1, 10))
 @example((-5, 0.0, -0.0))
 @example((0.0, -0.0, 5))
+@example((-0.0, 0.0, 0.0))
+@example((-0.0, -0.0, 0.0))
 @given(clamp_inputs())
 def test_clamp(input):
     lower, value, upper = input

--- a/hypothesis-python/tests/cover/test_float_utils.py
+++ b/hypothesis-python/tests/cover/test_float_utils.py
@@ -70,7 +70,9 @@ def test_next_float_equal(func, val):
 @example(float_kw(1, 4, smallest_nonzero_magnitude=4), -4)
 @example(float_kw(1, 4, smallest_nonzero_magnitude=4), -5)
 @example(float_kw(1, 4, smallest_nonzero_magnitude=4), -6)
-@example(float_kw(-5e-324, -0.0, smallest_nonzero_magnitude=5e-324), 3.0)
+@example(float_kw(-5e-324, -0.0), 3.0)
+@example(float_kw(0.0, 0.0), -0.0)
+@example(float_kw(-0.0, -0.0), 0.0)
 @given(float_kwargs(), st.floats())
 def test_float_clamper(kwargs, input_value):
     min_value = kwargs["min_value"]

--- a/hypothesis-python/tests/nocover/test_strategy_state.py
+++ b/hypothesis-python/tests/nocover/test_strategy_state.py
@@ -15,7 +15,7 @@ from random import Random
 from hypothesis import Verbosity, assume, settings
 from hypothesis.database import InMemoryExampleDatabase
 from hypothesis.internal.compat import PYPY
-from hypothesis.internal.floats import float_to_int, int_to_float, is_negative
+from hypothesis.internal.floats import clamp, float_to_int, int_to_float, is_negative
 from hypothesis.stateful import Bundle, RuleBasedStateMachine, rule
 from hypothesis.strategies import (
     binary,
@@ -35,18 +35,6 @@ from hypothesis.strategies import (
 )
 
 AVERAGE_LIST_LENGTH = 2
-
-
-def clamp(lower, value, upper):
-    """Given a value and optional lower/upper bounds, 'clamp' the value so that
-    it satisfies lower <= value <= upper."""
-    if (lower is not None) and (upper is not None) and (lower > upper):
-        raise ValueError(f"Cannot clamp with lower > upper: {lower!r} > {upper!r}")
-    if lower is not None:
-        value = max(lower, value)
-    if upper is not None:
-        value = min(value, upper)
-    return value
 
 
 class HypothesisSpec(RuleBasedStateMachine):

--- a/hypothesis-python/tests/pytest/test_collection_warning.py
+++ b/hypothesis-python/tests/pytest/test_collection_warning.py
@@ -1,0 +1,37 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import pytest
+
+pytest_plugins = "pytester"
+
+INI = """
+[pytest]
+norecursedirs = .svn tmp whatever*
+"""
+
+TEST_SCRIPT = """
+def test_noop():
+    pass
+"""
+
+
+@pytest.mark.skipif(int(pytest.__version__.split(".")[0]) < 7, reason="hook is new")
+def test_collection_warning(pytester):
+    pytester.mkdir(".hypothesis")
+    pytester.path.joinpath("pytest.ini").write_text(INI, encoding="utf-8")
+    pytester.path.joinpath("test_ok.py").write_text(TEST_SCRIPT, encoding="utf-8")
+    pytester.path.joinpath(".hypothesis/test_bad.py").write_text(
+        TEST_SCRIPT.replace("pass", "raise Exception"), encoding="utf-8"
+    )
+
+    result = pytester.runpytest_subprocess()
+    result.assert_outcomes(passed=1, warnings=1)
+    assert "Skipping collection of '.hypothesis'" in "\n".join(result.outlines)

